Boost: add full page caching

donnchawp opened the pull request:
A WIP snapshot of what this code looks like.

This Boost module will create cached files from requests and place them in wp-content/boost-cache/cache/
It avoids caching POST and "backend" requests like wp-admin/ajax/rest.
It uses the cookie as the cache key. Cache files are stored in separate directories so it would be possible to track where one user goes.

No garbage collection, no effort to delete cache files when a post is updated.

Fixes #35042

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-2rA-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
* Install Boost and apply this PR.
* Define the constant BOOST_CACHE = true in wp-config.php
* Visit the Boost settings page in wp-admin which will fire off the code that creates wp-content/advanced-cache.php and adds the WP_CACHE define to wp-config.php
* Look in wp-content/boost-cache/cache/ for new directories and files. `ls -lR` will be useful there as directories are deep.
* View the source of a page, scroll to the end, and you should see the text, "&lt;!-- cached -->" if the page was served from cache.

donnchawp committed:
Caching engine for Boost with debug logs

donnchawp committed:
changelog

donnchawp committed:
Rename this to match calls to this function.

donnchawp committed:
Add GET array to cache key and a filter on it.

donnchawp committed:
Log the actual cache file used to serve a page

thingalon commented:
Woah! Turning on the cache, it's fast :)

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,81 @@
+<?php
+
+class Boost_Cache {
+	public $key         = false;
+	public $request_uri = false;
```
IMHO `$key` ought to be private, as even child classes will need to use `->key()` to get it. `$request_uri` ought to be `protected`.

donnchawp replied:
You're right, thanks!

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,81 @@
+		add_action( 'wp', array( $this, 'sanitize' ), 0 );
+	}
+
+	public function sanitize() {
+		if ( ! $this->request_uri ) {
+			return;
+		}
+
+		$this->request_uri = esc_url_raw( wp_unslash( $this->request_uri ) );
+	}
```
🤔 Is this delayed until `wp_unslash` is available?

donnchawp replied:
Yeah, it runs on the "wp" action. It's possibly not needed, but in Super Cache I needed to keep a backup of it in case plugins messed around with it. I used it when storing meta information, and some checks.

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
```diff
@@ -0,0 +1,71 @@
+		return false;
+	}
+
+	private function createCacheDirectory( $path ) {
```
`snake_case` in PHP.

donnchawp replied:
Of course. Thanks.

thingalon reviewed (commented):
Exciting to see this in progress. I've thrown in a few mild comments, but don't want to get into fussing over all the little things until it's ready for review. Nice start! :)

donnchawp committed:
Restrict access to these variables, props @thingalon

donnchawp committed:
Fix the function name, props @thingalon

donnchawp committed:
Add REQUEST_URI to key. Not gonna use request path.

donnchawp committed:
Only cache text/html requests

donnchawp committed:
Split keys into cache and path keys and use md5 of path

donnchawp committed:
Add function comments.

donnchawp committed:
Move location of boost_cache_cacheable so it's an override.

donnchawp committed:
Add Module file for setup and UI

donnchawp committed:
Add Page_cache to Modules index

donnchawp committed:
Add "page-cache" to premium features. Not used yet.

donnchawp committed:
Remove error_log statements

donnchawp committed:
Write cache to a tmp file first.

This avoids other visitors requesting a possibly incomplete file while
it is being written to. The rename function can't be interrupted and
happens "instantly", in comparison.

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+	protected $request_uri = false;
+
+	public function __construct() {
+		$this->request_uri = isset( $_SERVER['REQUEST_URI'] )
+			? $_SERVER['REQUEST_URI'] // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.ValidatedSanitizedInput.MissingUnslash
+			: false;
```
I find the dataflow around `$this->request_uri` a little hard to follow. First, I want to check that the complexity is warranted, and if so recommend that we add a comment explaining the flow.

It looks like `Boost_Cache` is only instantiated (via its subclass) in `advanced-cache.php` at the moment. My analysis of the data-flow below is based on that premise, but it may be more complex if this class will ever be used from wp-admin or wp_cron when managing the cache.

Here is how I think this works:

1. As soon as `advanced-cache.php` runs, it grabs and stores the value in `$_SERVER['REQUEST_URI']` and stores it in a local variable.
2. Once the `wp` event fires, it replaces this variable with a sanitized version of itself.

That means that anyone that consumes this variable needs to be aware of when during WordPress' lifecycle it accesses it - otherwise it may get a sanitized or unsanitized version.

In fact, reading `file_cache.php`, it looks like `Boost_File_Cache::get()` will run _before_ sanitization, and `Boost_File_Cache::set()` will run _after_ sanitization, so if sanitization has any impact on the URL, then the path used when saving the cached page will not match the path used to access it.

If this complex dataflow is necessary, it needs to be clear in the code/comments as to why. Otherwise, we should simplify this.

donnchawp replied:
You're right that the get() and set() functions might get different request_uri variables. I think we can store the MD5 of the request_uri directly.  We're using the MD5 of the path now for the cache filename. After all, the first thing `calculate_path()` does is get the md5 of the `$request_uri`.

Later, if we want to show a listing of the cache pages, I think we will need to record the path. That path can be sanitized in the same way that WP Super Cache does, and that's possible before WordPress loads. That info can either go in the cache file in a JSON array, or a separate sidecar file alongside each cache file. It will be escaped before displaying too.

donnchawp replied:
After thinking about it more, we still need the request_uri - it's used to calculate the cache key and the path to the cache file, and it can be sanitized. I've copied the sanitization code from WP Super Cache.

thingalon replied:
Yeah, I agree that we need to bundle whatever sanitization methods we need in the Cache class so that it can self-sufficiently sanitize the string once and keep it. :)

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
There are two file naming standards used throughout the Boost codebase. Sadly. We are migrating from the old way (`class-boost-cache.php`), to naming files after the class they contain. This file should be `Boost_Cache.php` and it should be namespaced unless there's a good reason not to.

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+class Boost_Cache {
+	private $cache_key     = false;
+	private $path_key      = false;
+	protected $request_uri = false;
```
Can we please have docblocks for each variable describing what they do, and what type they are?

donnchawp replied:
Done. and I was able to remove $path_key since it's the same as $path in the file cache. If we have different cache backends we can put it back in, if required.

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+	 *
+	 * This is called on the wp action, because sanitization functions are not available before that.
+	 */
+	public function sanitize() {
```
The name `sanitize` doesn't make the purpose of the function clear. I'd aim for more descriptive names.

donnchawp replied:
That function has been removed, and replaced by `sanitize_request_uri()`.

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+			return false;
+		}
+
+		if ( ( isset( $_SERVER['REQUEST_METHOD'] ) && $_SERVER['REQUEST_METHOD'] === 'POST' ) ) {
```
Unnecessary double brackets.

donnchawp replied:
Thanks, removed!

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+	 *
+	 * @return string
+	 */
+	private function calculate_cache_key( $request_uri = '' ) {
```
It seems a bit strange to accept a `$request_uri` parameter here, given this function uses superglobals like `$_COOKIE` and `$_GET`, which are associated with the current request.

If you ever supply an alternate `$request_uri`, the cache key it generates will still use the cookies and parameters from the current request. Is there a situation where we might want that behaviour?

donnchawp replied:
You're right. I should pass those as optional parameters too. The idea being, that given a URL, we can pass it to this function to get the cache key MD5. That, along with the correct path, can be used to do admin tasks like deleting associated cache files when a post is edited.

donnchawp replied:
I overhauled the parameters, putting them in an $args array, and brought them all the way up to the get() and set() functions.

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+	 * @param string $request_uri (optional) The request uri to use to calculate the path key. Defaults to the current request uri.
+	 * @return string
+	 */
+	public function path_key( $request_uri = '' ) {
+		if ( $request_uri !== '' ) {
+			return md5( $request_uri );
+		}
+
+		if ( ! $this->path_key ) {
+			$this->path_key = md5( $this->request_uri );
+		}
+		return $this->path_key;
+	}
```
What is the difference between a `path_key` and a `cache_key`? I can't see this function used anywhere, so I don't understand its purpose.

donnchawp replied:
The path_key here is the md5 of the request_uri, as you can see. I forgot to use it in the file cache in calculate_path(), but I've just fixed that. I removed path_key, but then realised it should be used and put it back again.

We could have multiple cache files for a particular URL. One user could be completely anonymous, another may leave a comment, another will click a cookie banner, all resulting in different cookies.
With the path and cookie keys split up, it's simple to delete all the cache files for that URL in one go, if a new comment is made or the page is edited.

thingalon replied:
Ah, that makes a great deal of sense. Thank you! :)

I see you updated the comments to reflect this explanation too.

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+		return $this->path_key;
+	}
+
+	public function get() {
+		return false;
+	}
```
If this is about defining an interface for base classes to override, I recommend using an abstract function.

donnchawp replied:
Yes, I'll do that.

thingalon commented on projects/plugins/boost/app/modules/cache/cache.php:
```diff
@@ -0,0 +1,168 @@
+		return false;
+	}
+
+	public function set( $data ) {
+		echo esc_html( $data );
+		return false;
+	}
```
Won't `esc_html` escape all the HTML tags in the output? Also seems like a good case to use an abstract method.

donnchawp replied:
Same as for get(). I'll make it an abstract function.

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
File should be called `Boost_File_Cache`. I recommend namespacing it too. Yeah, it does mean that `advanced-cache.php` will need to reference it by `\Full\Name\Space\Boost_File_Cache`, but I think it's worth that little bit of mess for keeping the global namespace clear.

donnchawp replied:
Done. Thanks.

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
```diff
@@ -0,0 +1,84 @@
+
+require_once __DIR__ . '/cache.php';
+
+class Boost_File_Cache extends Boost_Cache {
```
The functions and members of this class are undocumented; please add docblocks.

donnchawp replied:
Done

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
```diff
@@ -0,0 +1,84 @@
+	private $path           = false;
+	private $cache_filename = false;
+
+	private function calculate_path( $request_uri ) {
+		$key  = md5( $request_uri );
+		$path = WP_CONTENT_DIR . DIRECTORY_SEPARATOR . 'boost-cache' . DIRECTORY_SEPARATOR . 'cache' . DIRECTORY_SEPARATOR;
+		for ( $i = 0; $i <= 5; $i++ ) {
+			$path .= substr( $key, $i, 1 ) . DIRECTORY_SEPARATOR;
+		}
+
+		return $path;
+	}
```
I notice that when it creates a directory for cache files, it only takes the request_uri into account, and MD5s it. It will generate a different MD5 hash to the full cache filename which will take the GET params and COOKIE into account. 

What's the benefit of generating a different MD5 hash for the directory structure?

donnchawp replied:
It allows us to do admin work on URLs more easily. If a page is edited, we can delete all the cached files for that page by getting the path to it. All the visitors who ever visited that page will have cached pages in that single directory.

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
```diff
@@ -0,0 +1,84 @@
+
+	private function cache_filename() {
+		if ( $this->cache_filename === false ) {
+			$this->cache_filename = $this->path() . $this->cache_key() . '.php';
```
If we're saving all cache files as `.php`, could it ever be possible for a user to inject a malicious comment or something which displays `<?php` tags in the page output, and then execute it by directly calling the relevant `.php` file from the cache directory?

Do we need to prepend a `<?php\ndie();\n` to every cached file for safety? Would it make sense to cache them with `.html` on the end?

donnchawp replied:
You're right. I was thinking of adding meta data like headers, and canonical URL to this file, but it could be added as a meta sidecar file, just as easily. In WP Super Cache, all cache files are .html while meta files are .php and do have a `die()` at the start of each file.

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
```diff
@@ -0,0 +1,84 @@
+		$tmp_filename = $this->cache_filename() . uniqid( wp_rand(), true ) . '.tmp';
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_file_put_contents
+		file_put_contents( $tmp_filename, $data );
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.rename_rename
+		rename( $tmp_filename, $this->cache_filename() );
```
How will `rename` behave if there is already a cache file there? Does it need to delete first?

donnchawp replied:
An existing file will be [overwritten](https://www.php.net/manual/en/function.rename.php). There's a good chance we'll have race conditions where a particularly popular page is generated more than once when someone leaves a comment. That particular problem is solved by the rebuild feature in WP Super Cache. One for later.
Renames (on UNIX anyway) are [atomic](https://www.php.net/manual/en/function.rename.php#2430). If multiple visitors are generating cache files, as soon as one is renamed to the final filename, everyone else gets to see a cached page. New updates to that file don't interrupt visitors reading from the file. They simply get the old file, which might be only a few milliseconds old.

thingalon replied:
Nice! Thanks for the explanation :)

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
```diff
@@ -0,0 +1,84 @@
+		die();
+	}
+
+	public function ob_start() {
+		if ( ! $this->is_cacheable() ) {
+			return false;
+		}
+
+		ob_start( array( $this, 'set' ) );
+	}
```
Does this belong in the base class?

donnchawp replied:
You're right. Done.

thingalon commented on projects/plugins/boost/app/modules/cache/file_cache.php:
```diff
@@ -0,0 +1,84 @@
+	}
+}
+
+$boost_cache = new Boost_File_Cache();
+if ( ! $boost_cache->get() ) {
+	$boost_cache->ob_start();
+}
```
I don't like putting procedural code at the bottom of files that define classes. It can be a surprise for anyone reading the class file that it does stuff immediately on load.

Would it make sense to create a function in `Boost_Cache` called something like `serve` which contains this `->get()` and `->ob_state()` logic, so that the advanced-cache.php file can look like this:

```
require_once "path to Boost_File_Cache.php";
( new Boost_File_Cache() )->serve();
```
?

That way advanced-cache can be kept as simple as possible, but these class declaration files can avoid including procedural code at the bottom :)

donnchawp replied:
Great idea, thanks. Done.

thingalon commented on projects/plugins/boost/app/modules/cache/page_cache.php:
```diff
@@ -0,0 +1,84 @@
+	 */
+	public function setup() {
+		if ( is_admin() ) {
+			require_once ABSPATH . '/wp-admin/includes/file.php';
+			WP_Filesystem();
+			$error = $this->create_advanced_cache();
+			if ( $error === true ) {
+				$this->add_wp_cache_define();
+			}
+			// do something with $error
```
Should this run every time an admin page loads?

donnchawp replied:
No, it should really only run on the Jetpack Boost page. I'll add a page check.

thingalon replied:
👍 Ok.

thingalon commented on projects/plugins/boost/app/modules/cache/page_cache.php:
```diff
@@ -0,0 +1,84 @@
+			require_once ABSPATH . '/wp-admin/includes/file.php';
+			WP_Filesystem();
+			$error = $this->create_advanced_cache();
+			if ( $error === true ) {
```
`if ( $error === true ) { do more stuff }`

Maybe it should be `$result` :)

donnchawp replied:
Oh, yes. Done

thingalon commented on projects/plugins/boost/app/modules/cache/page_cache.php:
```diff
@@ -0,0 +1,84 @@
+// Boost Cache Plugin 0.1
+require_once( ABSPATH . \'/wp-content/plugins/boost/app/modules/cache/file_cache.php\' );
+';
+			$wp_filesystem->put_contents( $advanced_cache_filename, $contents, FS_CHMOD_FILE );
```
Might be wise to check the response of this.

donnchawp replied:
Added a check on wp-config.php too.

thingalon commented on projects/plugins/boost/app/modules/cache/page_cache.php:
```diff
@@ -0,0 +1,84 @@
+	private function add_wp_cache_define() {
+		global $wp_filesystem;
+
+		$content = $wp_filesystem->get_contents( ABSPATH . 'wp-config.php' );
+		if ( strpos( $content, 'define( \'WP_CACHE\', true );' ) !== false ) {
+			return; // WP_CACHE already added.
+		}
```
Would it be worth checking `defined( 'WP_CACHE' )` too?

donnchawp replied:
Good suggestion, but it might mean that WP_CACHE is defined in an mu-plugin or regular plugin. In which case, advanced-cache.php wouldn't be loaded. The only place it could be defined is in the config file or the code file, wp-settings.php.
Anywhere else, there should be a defined() check before defining it.

thingalon replied:
Ok, that seems reasonable. Thanks! 👍

thingalon reviewed (changes requested):
It's nice to see a working cache in Boost. I have suggestions, recommendations and questions - inline below.

donnchawp committed:
Add sanitize_request_uri to sanitize the request_uri and comments

donnchawp committed:
Sanitize the request_uri, add comments

donnchawp committed:
Removed brackets, they're not needed

donnchawp committed:
Fix parse_url returning "" for the path, and add a slash

donnchawp committed:
Add path_key back, and use it in file cache

donnchawp committed:
Make these abstract, the file cache does heavy lifting

donnchawp committed:
Rename and add namespace

donnchawp committed:
Launch cache engine from constructor. No global vars.

donnchawp committed:
Move page serving code into serve()

donnchawp committed:
Move ob_start code into base class.

donnchawp committed:
Store cache files as html, not php. Safer.

donnchawp committed:
Only check for advanced-cache on the JB settings page

donnchawp committed:
Check that config files were written correctly

donnchawp committed:
Rename, part 1. Case insensitivity ..

donnchawp committed:
Rename to match class name case.

donnchawp committed:
Add optional $args, and surface them to public functions

donnchawp committed:
Add comments

donnchawp committed:
Remove the $path variable. Not used.

donnchawp committed:
After creating cache file, serve it through get()

donnchawp committed:
Renaming on case insensitive fs. p1

donnchawp committed:
Renaming on a case insensitive fs, p2

Commit:
Merge branch 'trunk' into add/cache_file

Commit:
Fix broken link in url

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -0,0 +1,194 @@
+			$request_uri .= '/';
+		}
+
+		$request_uri = $this->deep_replace(
+			array( '..', '\\', 'index.php' ),
```
I may have missed something here, but I don't understand the purpose of using `deep_replace` in this situation. As I understand it, `deep_replace` re-runs `str_replace` over and over so that strings which "collapse" into other meaningful strings also get replaced.

But we're replacing every `..`, `\` and `index.php` with `''`, I don't see how one str_replace could every lead to needing a second one without coming up silly scenarios like `indindex.phpex.php` which would collapse twice over. :)

What am I missing?

donnchawp replied:
I could probably leave `index.php` out of that list (as it'll be removed by parse_url anyway), but we need the deep_replace to make sure ".." doesn't end up in the final path, which would allow an attacker to perform a directory traversal. "\" doesn't belong in URLs either, and might be used to escape a character.

donnchawp replied:
I removed the entire deep_replace() code and renamed the function "normalize_request_uri", with dire comments about using the variable. :)

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -0,0 +1,194 @@
+	 * @return string
+	 */
+	public function cache_key( $args = array() ) {
+		$defaults = array(
+			'request_uri' => $this->request_uri,
+			'cookies'     => $_COOKIE, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			'get'         => $_GET, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Recommended
+		);
+		$args     = array_merge( $defaults, $args );
+
+		if ( $args['request_uri'] === '' ) {
+			$args['request_uri'] = $this->request_uri;
+		} else {
+			$args['request_uri'] = $this->sanitize_request_uri( $args['request_uri'] );
+		}
```
The only way `$args['request_uri']` would be `''` is if the caller explicitly passed in `array( 'request_uri' => '' );` as the default is the pre-sanitized current request URI. Not sure this is the intended behaviour.

Perhaps before merging with the defaults, you can say `if ( isset( $args['request_uri'] ) ) { ... sanitize ... }` ?

donnchawp replied:
I wanted to catch when cache_key might be accidentally called with a blank request_uri, say, when WP_CLI is used, even though caching should be disabled. The `array_merge` handles where request_uri isn't passed to the function at all.

However, thinking about it again. I think it's OK if request_uri is blank. The cache_key only defines the base name of the filename. It doesn't include the path, so any cache file generated will be in the directory allocated to the current URL. It's still differentiated by cookies and get parameters.

I'll do as you suggested. Thanks.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,122 @@
+		if ( ! $this->get() ) {
+			$this->ob_start();
+		}
+	}
```
Does this belong in `Boost_File_Cache` or in `Boost_Cache`? Seems like core cache behaviour to me.

donnchawp replied:
Yeah, you're right. Moved it over.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param array $args - an array containing the request_uri, cookies array, and get array representing the request.
+	 * @return string - The cache path + filename for the given request.
+	 */
+	private function cache_filename( $args = array() ) {
+		$defaults = array(
+			'request_uri' => $this->request_uri,
+			'cookies'     => $_COOKIE, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			'get'         => $_GET, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Recommended
+		);
+		$args     = array_merge( $defaults, $args );
+
+		return $this->path( $args['request_uri'] ) . $this->cache_key( $args ) . '.html';
+	}
```
This passes the $args object down to `$this->cache_key`, which will receive the arguments and then re-sanitize them. `$this->request_uri` will already be sanitized on class construction. Is there any danger to sanitizing the request_uri twice, or will it always result in the same value?

donnchawp replied:
It's safe to normalize the value twice, definitely since I removed the deep_replace. This should be ok, but I'll add a check here, if it's a different request_uri, then normalize it.

thingalon commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -0,0 +1,100 @@
+		global $wp_filesystem;
+
+		$content = $wp_filesystem->get_contents( ABSPATH . 'wp-config.php' );
+		if ( strpos( $content, 'define( \'WP_CACHE\', true );' ) !== false ) {
+			return; // WP_CACHE already added.
+		}
```
This check requires that `WP_CACHE` has been added using our coding standards, with spaces inside the brackets - and that the value is `true`. This won't identify:

```
define('WP_CACHE',true);
define( 'WP_CACHE', false );
```

Perhaps a regexp is the way to go here?

Also it may hit a false positive if there is a commented-out `WP_CACHE` define in the file - but detecting that is getting ridiculous. Perhaps it should check `defined( 'WP_CACHE' )` to rule out commented out lines, and use a regexp to detect anything that looks like any kind of `WP_CACHE` setter?

e.g.:
```
$is_defined = preg_match( '#define\s*\(\s*(["'])WP_CACHE\1\s*,[^\)]+\)#', $content ) && defined( 'WP_CACHE' );
$is_truthy = WP_CACHE;
```

?

dilirity replied:
I've seen this definition as well (on JN):

```
define( 'WP_CACHE', 'true' );
```

Value wrapped in quotation marks.

dilirity replied:
Looking at this again - doesn't WP have some function to check these?

donnchawp replied:
> Looking at this again - doesn't WP have some function to check these?

Well, it has tools to add blocks of text to the .htaccess file, but not to wp-config.php, but ithemes security uses something similar to add their own blocks of code to wp-config.php, while messing up WP Super Cache's WP_CACHE define.
There's no special commands for modifying the config file.

donnchawp replied:
I think a regex is the way to go. I added one that includes the string true with ', ", or bare.

thingalon reviewed (changes requested):
Nearly there! Thanks for adding my last round of recommendations. I've spotted more concerns and questions. :)

donnchawp committed:
No need to remove index.php, parse_url already did.

donnchawp committed:
Makes more sense to have this in the base class

donnchawp committed:
Sanitize the request_uri. It OK to be blank here.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+		$path = WP_CONTENT_DIR . DIRECTORY_SEPARATOR . 'boost-cache' . DIRECTORY_SEPARATOR . 'cache' . DIRECTORY_SEPARATOR;
+		for ( $i = 0; $i <= 5; $i++ ) {
+			$path .= substr( $key, $i, 1 ) . DIRECTORY_SEPARATOR;
+		}
```
I don't understand what this does.

Could you add a comment explaining why it's there and what it does?

donnchawp replied:
I added a comment, and changed that directory structure so it uses 2 characters per directory instead of 1. That gives us max 256 directories per directory, which will be plenty of storage, for hopefully the largest sites.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+		}
+
+		$key  = $this->path_key( $request_uri );
+		$path = WP_CONTENT_DIR . DIRECTORY_SEPARATOR . 'boost-cache' . DIRECTORY_SEPARATOR . 'cache' . DIRECTORY_SEPARATOR;
```
Shouldn't the cache folder be `wp-content/cache/boost`? Or is `cache` reserved for super cache 😅

donnchawp replied:
I think we should keep everything in a separate directory. We can put the config file in the boost-cache directory too. Some site owners don't like wp-content being writable. If you want to write to a file in that directory (make a temp file first, then rename it) then you need write access.

dilirity replied:
> If you want to write to a file in that directory (make a temp file first, then rename it) then you need write access.

I had no idea! Sneaky sneaky.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+		);
+		$args     = array_merge( $defaults, $args );
+
+		if ( file_exists( $this->cache_filename( $args ) ) ) {
```
I think you can just pass `$args` to `cache_filename` as that handles defaults.

donnchawp replied:
You're right, thanks!

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+	private function create_cache_directory( $path ) {
+		if ( ! is_dir( $path ) ) {
+			// phpcs:ignore WordPress.WP.AlternativeFunctions.dir_mkdir_dirname, WordPress.WP.AlternativeFunctions.file_system_operations_mkdir
+			mkdir( $path, 0755, true );
```
Should we maybe check the result of `mkdir`? It could return `false`, though I'm not sure under what circumstances.

Maybe we should do a check when enabling the module, if where we're going to be writing is accessible.

donnchawp replied:
Definitely. As you said elsewhere, the module needs to return an error status to the UI. The setup code uses WP_Filesystem, but caching uses bare PHP file_put_contents, etc. So, the cache files will need to be writable. They might not be if WP_Filesystem uses SFTP.

I think we need:
* is_writeable check.
* use bare file_get/put_contents during setup.

donnchawp replied:
And that's done. Needs testing.

donnchawp replied:
Added a check on mkdir(). We'll need to attach it to a UI next.

thingalon replied:
Nice catch. For now it's ok to leave a TODO about logging stuff for the front-end because we don't have a front-end to log to just yet :)

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+			return false;
+		}
+
+		$args = array(
+			'request_uri' => $this->request_uri,
+			'cookies'     => $_COOKIE, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			'get'         => $_GET, // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Recommended
+		);
+
+		$cache_filename = $this->cache_filename( $args );
```
Since `$args` is optional, there's no need to pass it to `cache_filename`.

donnchawp replied:
Yep, thanks!

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+		file_put_contents( $tmp_filename, $data );
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.rename_rename
+		rename( $tmp_filename, $cache_filename );
+		$this->get( $args );
```
Can we avoid the write and read happening back to back by outputting the `$data` directly?

Also - can we directly write to `$cache_filename` instead of a temp file first?

donnchawp replied:
Yes, writing it directly was the way I had it working first, but then I changed to using get(), but then I noticed that wasn't the right way to do it. The code should simply return the output buffer, and it will be sent to the browser. 

Writing to a temp file first allows a busy site to serve the cached file to other visitors while a new file is being created. The rename() is atomic, and overwrites the old cache file. This is in the event that multiple people are creating a cache file for the same page at the exact same time.
WP Super Cache has a "rebuild" feature that serves the old cache page while a new one is being built, a feature we should bring here eventually as it helps a ton.

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -0,0 +1,100 @@
+	 */
+	public function setup() {
+		if ( is_admin() && isset( $_GET['page'] ) && $_GET['page'] === 'jetpack-boost' ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			require_once ABSPATH . '/wp-admin/includes/file.php';
+			WP_Filesystem();
+			$result = $this->create_advanced_cache();
+			if ( $result === true ) {
+				$result = $this->add_wp_cache_define();
+				if ( $result !== true ) {
+					return $result;
+				}
+			} else {
+				return $result;
+			}
+
+			return true;
```
From what I could find, we don't handle returns for a module setup.

Since this is the first module that needs a setup (I think), maybe we need to expand it in a way? Like check if folders are writable, create files, add defines, etc.

Though I don't think we need to do it now.

donnchawp replied:
yep. Definitely needs to be done.

thingalon replied:
Can we turn this into a ticket on the board / TODO list for Boost cache?

donnchawp replied:
Done. Added in [#35215](https://github.com/Automattic/jetpack/issues/35215).

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -0,0 +1,100 @@
+	 * Returns true if the files were setup correctly, or WP_Error if there was a problem.
+	 * @return bool|WP_Error
+	 */
+	private function create_advanced_cache() {
+		global $wp_filesystem;
+
+		$advanced_cache_filename = ABSPATH . 'wp-content/advanced-cache.php';
+		if ( $wp_filesystem->exists( $advanced_cache_filename ) ) {
+			$content = $wp_filesystem->get_contents( $advanced_cache_filename );
+			if ( strpos( $content, 'Boost Cache Plugin 0.1' ) !== false ) {
+				return true;
+			} else {
+				return new \WP_Error( 'advanced-cache.php exists but is not the correct file' );
+			}
+		} else {
+			$contents = '<?php
+// Boost Cache Plugin 0.1
+require_once( ABSPATH . \'/wp-content/plugins/boost/app/modules/cache/Boost_File_Cache.php\' );
+
+( new Automattic\Jetpack_Boost\Modules\Page_Cache\Boost_File_Cache() )->serve();
+';
+
+			$result = $wp_filesystem->put_contents( $advanced_cache_filename, $contents, FS_CHMOD_FILE );
+			if ( $result === false ) {
+				return new \WP_Error( 'Could not write to advanced-cache.php' );
+			}
+		}
+
+		return true;
+	}
```
Maybe we could avoid adding the version check in there if we just include a file from Boost, that does the require and `serve()`?

I think we should simplify the contents of advanced-cache.php as much as possible, as we might run into issues if we decide to change up structures or reorganize the file structure.

donnchawp replied:
The version check here, is to just make sure the advanced-cache.php belongs to Boost. We don't want to overwrite a file created by another plugin.
I had to update advanced-cache.php in WP Super Cache once, a long time ago, and I checked for a string to see if it had updated.

thingalon replied:
Yeah, the version comment has to go directly in `advanced-cache.php` to have any value here.

As for the rest of it, we could put it in a php file of its own and require that from advanced-cache. That could shorten the advanced-cache from 3 lines to 2. Personally I'm not too fussed either way. :)

dilirity replied:
> Yeah, the version comment has to go directly in advanced-cache.php to have any value here.

Why is that necessary? I'm missing the point of adding a lot of things to the file.

donnchawp replied:
@dilirity - Mainly, it's to allow us to update the advanced-cache.php more easily in the future, if an update is necessary.

dilirity replied:
Oh I see, doh :D Compare if it's outdated. Got it.

dilirity commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	public static function is_available() {
+		if ( ! defined( 'BOOST_CACHE' ) ) {
+			return false;
+		}
```
Nice!

dilirity reviewed (changes requested):
Wow great work on this <3

TTFB with cache: 9ms
TTFB w/o cache: 190ms

:D

I tried disabling the cache by removing the constant, but I forgot that WP is bypassed, so the logic that checks if the module is active is not loaded.

I deleted `advanced-cache.php` in order to disable it. Also, the testing instructions are showing `""` instead of the comment I imagine you meant to write `<-- cached -->`.

I've added some notes and questions. Oh and changelogger isn't happy.

donnchawp committed:
It's safer to exclude JSON from caching then check for html

donnchawp committed:
The output buffer should return the data for display

donnchawp committed:
Normalize, not sanitize the REQUEST_URI for now.

donnchawp committed:
Only normalize if the URI is different to the internal one

donnchawp committed:
Remove DIRECTORY_SEPARATOR, "/" works fine on Windows too.

donnchawp committed:
Less chance of collisions with 2 char directories

donnchawp committed:
Comment explaining cache directory structure

donnchawp committed:
Optional params are handled by cache_filename()

donnchawp committed:
If not cacheable then return the page to browser

donnchawp committed:
If no output buffer, then don't cache anything. ie. redirects

donnchawp committed:
Add 404 and feed exclusions to caching

donnchawp committed:
vscode was complaining this wasn't defined, so constant() it is.

donnchawp committed:
Use bare PHP to write setup files

donnchawp committed:
Check that mkdir() succeeded

donnchawp committed:
Give this temp file a random filename.

donnchawp committed:
Use a regex to detect WP_CACHE

donnchawp committed:
Update versions

donnchawp committed:
Merge branch 'trunk' into add/cache_file

thingalon reviewed (dismissed):
Thanks for iterating on this so much @donnchawp. In my opinion, this is ready to be merged, and we can build on it from there. 

There's plenty more to do (such as a logging system) but we can burn those bridges when we get to them. This is a foundation we can build on.

@dilirity - I'll leave your request for changes as-is, so you can re-review and see if you're cool for a 👍 too

haqadn commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -0,0 +1,194 @@
+			return false;
+		}
+
+		if ( isset( $_SERVER['REQUEST_METHOD'] ) && $_SERVER['REQUEST_METHOD'] === 'POST' ) {
```
Shouldn't we cache only GET requests and ignore everything(PUT, PATCH, DELETE, etc.) else? I think it would be more accurate to only allow GET specifically.
```suggestion
		if ( isset( $_SERVER['REQUEST_METHOD'] ) && $_SERVER['REQUEST_METHOD'] !== 'GET' ) {
```

donnchawp replied:
Thanks, you're right. Applied the change!

haqadn commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+	 * @param string $data - The output buffer to save to the cache file.
+	 * @return bool - false if page was not cacheable.
+	 */
+	public function set( $data ) {
```
Just a minor suggestion.
IMHO, it returning the cached data can cause unintentional bugs. If the call depends on this for output and we end up updating this function to return false for more cases than `strlen( $data ) === 0`, we may end up in situations where we are returning blank pages where we could send the data.

In short, I think this should return boolean true or false. The caller would use this only to set the cache, but output it independently.

donnchawp replied:
Hmm, I see what you mean. The set function is used as the callback for ob_start(), but maybe the callback should be "ob_callback" and it should call set() to record the output buffer?

The output buffer callback returns the $data so that PHP sends it to the browser, but you're right. set() should return a bool to signal that there was a problem writing the cache file.

haqadn commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -0,0 +1,107 @@
+	 */
+	private function add_wp_cache_define() {
+		$content = file_get_contents( ABSPATH . 'wp-config.php' );
+		if ( preg_match( '#define\s*\(\s*[\'"]WP_CACHE[\'"]\s*,\s*(?:\'true\'|\"true\"|true)\s*\);#', $content ) === 1 ) {
+			return; // WP_CACHE already added.
+		}
+		$content = str_replace(
+			'<?php',
+			'<?php
+define( \'WP_CACHE\', true );',
+			$content
+		);
```
This will define `WP_CACHE` twice if `WP_CACHE` is anything other than true, i.e. `false`.

donnchawp replied:
That's "true". I'll remove the 'true' part. It's true that it was checking against too much text.

dilirity replied:
True, true.

thingalon replied:
Nice, I think this was a good move.

haqadn commented:
I appreciate that the code is really clean. It was super easy to understand.

haqadn commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,113 @@
+
+namespace Automattic\Jetpack_Boost\Modules\Page_Cache;
+
+require_once __DIR__ . '/Boost_Cache.php';
```
I assume we have this require becase we want this file to load even when the plugin and as a result the composer autoload file isn't loaded. In this case, it might be useful to add a comment so we accidentally do not remove the line.

donnchawp replied:
Yeah, that would be an excellent idea. Thanks.

donnchawp committed:
Update projects/plugins/boost/app/modules/cache/Boost_Cache.php

Exclude everything but GET requests.

Co-authored-by: Adnan Haque <3737780+haqadn@users.noreply.github.com>

donnchawp committed:
Add comments explaining how these are loaded.

donnchawp committed:
Check for WP_CACHE more carefully.

donnchawp committed:
Separate set() into ob_callback() and set()

donnchawp committed:
Merge branch 'add/cache_file' of github.com:Automattic/jetpack into add/cache_file

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_File_Cache.php:
```diff
@@ -0,0 +1,123 @@
+
+		$tmp_filename = $cache_filename . uniqid( wp_rand(), true ) . '.tmp';
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_file_put_contents
+		if ( ! file_put_contents( $tmp_filename, $buffer ) ) {
```
> This function may return Boolean false, but may also return a non-Boolean value which evaluates to false. Please read the section on [Booleans](https://www.php.net/manual/en/language.types.boolean.php) for more information. Use [the === operator](https://www.php.net/manual/en/language.operators.comparison.php) for testing the return value of this function.

Just to be safe, let's do a value/type check here.

donnchawp replied:
Thanks. I use file_put_contents in the is_writeable check too, so I fixed the comparison there too.

thingalon replied:
Nice catch @dilirity

dilirity reviewed (changes requested):
Small thing.

Other than that, I think we're good to go.

donnchawp committed:
Need to explicitly check the return value from here.

thingalon reviewed (approved):
All feedback points seem to have been addressed. Thanks @dilirity and @haqadn for reviewing it in depth too. Feeling pretty good about where it's at now.
